include/ocpp/v201/device_model.hpp

```diff
@@ -168,46 +127,118 @@ class DeviceModel {
+    virtual std::vector<ReportData> get_base_report_data(const ReportBaseEnum& report_base) = 0;
 
     /// \brief Gets the ReportData for the specifed filter \p component_variables and \p
     /// component_criteria
     /// \param report_base
     /// \param component_variables
     /// \param component_criteria
     /// \return
-    std::vector<ReportData>
-    get_custom_report_data(const std::optional<std::vector<ComponentVariable>>& component_variables = std::nullopt,
-                           const std::optional<std::vector<ComponentCriterionEnum>>& component_criteria = std::nullopt);
+    virtual std::vector<ReportData> get_custom_report_data(
+        const std::optional<std::vector<ComponentVariable>>& component_variables = std::nullopt,
+        const std::optional<std::vector<ComponentCriterionEnum>>& component_criteria = std::nullopt) = 0;
 
     /// \brief Check data integrity of the device model provided by the device model data storage:
     /// For "required" variables, assert values exist. Checks might be extended in the future.
-    void check_integrity(const std::map<int32_t, int32_t>& evse_connector_structure);
+    virtual void check_integrity(const std::map<int32_t, int32_t>& evse_connector_structure) = 0;
+
+private:
+    /// \brief Private helper method that does some checks with the device model representation in memory to evaluate if
+    /// a value for the given parameters can be requested. If it can be requested it will be retrieved from the device
+    /// model storage and the given \p value will be set to the value that was retrieved
+    /// \param component_id
+    /// \param variable_id
+    /// \param attribute_enum
+    /// \param value string reference to value: will be set to requested value if value is present
+    /// \param allow_write_only true to allow a writeOnly value to be read.
+    /// \return GetVariableStatusEnum that indicates the result of the request
+    virtual GetVariableStatusEnum request_value_internal(const Component& component_id, const Variable& variable_id,
+                                                         const AttributeEnum& attribute_enum, std::string& value,
+                                                         bool allow_write_only) = 0;
+};
+
+/// \brief This class manages access to the device model representation and to the device model storage and provides
+/// functionality to support the use cases defined in the functional block Provisioning
+class DeviceModel : public DeviceModelInterface {
+
+private:
+    DeviceModelMap device_model;
+    std::unique_ptr<DeviceModelStorage> storage;
+
+    GetVariableStatusEnum request_value_internal(const Component& component_id, const Variable& variable_id,
+                                                 const AttributeEnum& attribute_enum, std::string& value,
+                                                 bool allow_write_only) override;
```

Comment:
Depending on whether this `DeviceModel` class should be derived from again or not, it might be better to use `final` instead of `override` here and following.
Furthermore, the `DeviceModel` implementation could be moved into a separate header file if implementation details should not leak into the `DeviceModelInterface`

Reply:
Currently I don't think there is a reason to derive from `DeviceModel` again but I also don't see a reason to prohibit it by using final.

As for the interface. The only reason we are doing that is to be able to mock it. Not really to not leak implementation detail so that is why we have been using the interfaces in the same file as the header.